File: examples/python-examples/GlobalRegistration.dox.py
import meshlib.mrmeshpy as mrmeshpy
import sys


def print_stats(icp):
    num_active_pairs = icp.getNumActivePairs()
    print(f"Samples: {icp.getNumSamples()}")
    print(f"Active point pairs: {num_active_pairs}")

    if num_active_pairs > 0:
        p2pt_metric = icp.getMeanSqDistToPoint()
        p2pt_inaccuracy = icp.getMeanSqDistToPoint(p2pt_metric)
        print(f"RMS point-to-point distance: {p2pt_metric} ± {p2pt_inaccuracy}")

        p2pl_metric = icp.getMeanSqDistToPlane()
        p2pl_inaccuracy = icp.getMeanSqDistToPlane(p2pl_metric)
        print(f"RMS point-to-plane distance: {p2pl_metric} ± {p2pl_inaccuracy}")


def main(argv):
    if len(argv) < 4:
        print(f"Usage for script: {argv[0]} INPUT1 INPUT2 [INPUTS...] OUTPUT")
        return
    file_args = argv[1:]

    # the global registration can be applied to meshes and point clouds
    # to simplify the sample app, we will work with point clouds only
    input_clouds_num = len(file_args) - 1
    inputs = []
    # as ICP and MultiwayICP classes accept both meshes and point clouds,
    # the input data must be converted to special wrapper objects
    # NB: the wrapper objects hold *references* to the source data, NOT their copies
    max_bbox = None
    for i in range(input_clouds_num):
        points = mrmeshpy.loadPoints(file_args[i])
        # you may also set an affine transformation for each input as a second argument
        inputs.append(mrmeshpy.MeshOrPointsXf(points, mrmeshpy.AffineXf3f()))

        bbox = points.getBoundingBox()
        if not max_bbox or bbox.volume() > max_bbox.volume():
            max_bbox = bbox

    # you can set various parameters for the global registration; see the documentation for more info
    sampling_params = mrmeshpy.MultiwayICPSamplingParameters()
    # set sampling voxel size
    sampling_params.samplingVoxelSize = max_bbox.diagonal() * 0.03

    icp = mrmeshpy.MultiwayICP(mrmeshpy.Vector_MeshOrPointsXf_ObjId(inputs), sampling_params)
    icp.setParams(mrmeshpy.ICPProperties())

    # gather statistics
    icp.updateAllPointPairs()
    print_stats(icp)

    print("Calculating transformations...")
    xfs = icp.calculateTransformations()
    print_stats(icp)

    output = mrmeshpy.PointCloud()
    for i in range(input_clouds_num):
        xf = xfs.vec_[i]
        for point in inputs[i].obj.points():
            output.addPoint(xf(point))

    mrmeshpy.PointsSave.toAnySupportedFormat(output, file_args[-1])


main(sys.argv)

File: examples/cpp-examples/GlobalRegistration.dox.cpp
#include <MRMesh/MRBox.h>
#include <MRMesh/MRMultiwayICP.h>
#include <MRMesh/MRPointCloud.h>
#include <MRMesh/MRPointsLoad.h>
#include <MRMesh/MRPointsSave.h>

#include <iostream>

void printStats( const MR::MultiwayICP& icp )
{
    std::cout << "Samples: " << icp.getNumSamples() << std::endl
              << "Active point pairs: " << icp.getNumActivePairs() << std::endl;
    if ( icp.getNumActivePairs() > 0 )
    {
        const auto p2ptMetric = icp.getMeanSqDistToPoint();
        const auto p2plMetric = icp.getMeanSqDistToPlane();
        std::cout << "RMS point-to-point distance: " << p2ptMetric << " ± " << icp.getMeanSqDistToPoint( p2ptMetric ) << std::endl
                  << "RMS point-to-plane distance: " << p2plMetric << " ± " << icp.getMeanSqDistToPlane( p2plMetric ) << std::endl;
    }
}

int main( int argc, char* argv[] )
{
    if ( argc < 4 )
    {
        std::cerr << "Usage: " << argv[0] << " INPUT1 INPUT2 [INPUTS...] OUTPUT" << std::endl;
        return EXIT_FAILURE;
    }

    // the global registration can be applied to meshes and point clouds
    // to simplify the sample app, we will work with point clouds only
    std::vector<MR::PointCloud> inputs;
    // as ICP and MultiwayICP classes accept both meshes and point clouds,
    // the input data must be converted to special wrapper objects
    // NB: the wrapper objects hold *references* to the source data, NOT their copies
    MR::ICPObjects objects;
    MR::Box3f maxBBox;
    for ( auto i = 1; i < argc - 1; ++i )
    {
        auto pointCloud = MR::PointsLoad::fromAnySupportedFormat( argv[i] );
        if ( !pointCloud )
        {
            std::cerr << "Failed to load point cloud: " << pointCloud.error() << std::endl;
            return EXIT_FAILURE;
        }

        auto bbox = pointCloud->computeBoundingBox();
        if ( !maxBBox.valid() || bbox.volume() > maxBBox.volume() )
            maxBBox = bbox;

        inputs.emplace_back( std::move( *pointCloud ) );
        // you may also set an affine transformation for each input as a second argument
        objects.emplace_back( inputs.back() );
    }

    // you can set various parameters for the global registration; see the documentation for more info
    MR::MultiwayICP icp( objects, {
        // set sampling voxel size
        .samplingVoxelSize = maxBBox.diagonal() * 0.03f,
    } );

    icp.setParams( {} );

    // gather statistics
    icp.updateAllPointPairs();
    printStats( icp );

    std::cout << "Calculating transformations..." << std::endl;
    auto xfs = icp.calculateTransformations();
    printStats( icp );

    MR::PointCloud output;
    for ( auto i = MR::ObjId( 0 ); i < inputs.size(); ++i )
    {
        const auto& input = inputs[i];
        const auto& xf = xfs[i];
        for ( const auto& point : input.points )
            output.points.emplace_back( xf( point ) );
    }

    auto res = MR::PointsSave::toAnySupportedFormat( output, argv[argc - 1] );
    if ( !res )
    {
        std::cerr << "Failed to save point cloud: " << res.error() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}